internal/services/authorization/role_definition_data_source.go

```diff
@@ -4,178 +4,251 @@
+			if config.Name != "" {
+				// Accounting for eventual consistency
+				deadline, ok := ctx.Deadline()
+				if !ok {
+					return fmt.Errorf("internal error: context had no deadline")
+				}
+				err := pluginsdk.Retry(time.Until(deadline), func() *pluginsdk.RetryError {
+					roleDefinitions, err := client.List(ctx, commonids.NewScopeID(config.Scope), roledefinitions.ListOperationOptions{
+						Filter: pointer.To(fmt.Sprintf("roleName eq '%s'", config.Name)),
+					})
+					if err != nil {
+						return pluginsdk.NonRetryableError(fmt.Errorf("loading Role Definition List: %+v", err))
+					}
+					if roleDefinitions.Model == nil {
+						return pluginsdk.RetryableError(fmt.Errorf("loading Role Definition List: `Model` was nil"))
+					}
+					if len(*roleDefinitions.Model) != 1 {
+						return pluginsdk.RetryableError(fmt.Errorf("loading Role Definition List: could not find role '%s'", config.Name))
+					}
+					if (*roleDefinitions.Model)[0].Id == nil {
+						return pluginsdk.NonRetryableError(fmt.Errorf("loading Role Definition List: values[0].ID is nil '%s'", config.Name))
+					}
+
+					id = roledefinitions.NewScopedRoleDefinitionID(config.Scope, *(*roleDefinitions.Model)[0].Name)
+					resp, err := client.Get(ctx, id)
+					if err != nil {
+						return pluginsdk.NonRetryableError(fmt.Errorf("getting Role Definition by ID %s: %+v", defId, err))
+					}
+					if resp.Model == nil {
+						return pluginsdk.NonRetryableError(fmt.Errorf("retirving %s: `Model` was nil", id))
+					}
+					role = *resp.Model
+					return nil
+				})
+				if err != nil {
+					return err
+				}
+			} else {
+				var err error
+				id = roledefinitions.NewScopedRoleDefinitionID(config.Scope, defId)
+				resp, err := client.Get(ctx, id)
+				if err != nil {
+					return fmt.Errorf("retirving %s: %+v", id, err)
+				}
+				if resp.Model == nil {
+					return fmt.Errorf("retirving %s: `Model` was nil", id)
+				}
+				role = *resp.Model
 			}
 
-			defId = *roleDefinitions.Values()[0].ID
-			role, err = client.GetByID(ctx, defId)
-			if err != nil {
-				return pluginsdk.NonRetryableError(fmt.Errorf("getting Role Definition by ID %s: %+v", defId, err))
+			state := ArmRoleDefinitionDataSourceModel{
+				Scope:            id.Scope,
+				RoleDefinitionId: id.RoleDefinitionId,
 			}
-			return nil
-		})
-		if err != nil {
-			return err
-		}
-	} else {
-		var err error
-		role, err = client.Get(ctx, scope, defId)
-		if err != nil {
-			return fmt.Errorf("loading Role Definition: %+v", err)
-		}
+
+			if props := role.Properties; props != nil {
+				state.Name = pointer.From(props.RoleName)
+				state.Type = pointer.From(props.Type)
+				state.Description = pointer.From(props.Description)
+				state.Permissions = flattenDataSourceRoleDefinitionPermissions(props.Permissions)
+				state.AssignableScopes = pointer.From(props.AssignableScopes)
+			}
+
+			// if config.Scope == "", id.ID() will starts with "//", which is different with the previous id
```

Comment:
In which case this will be an empty string? If that is a valid case, then it seems to be a bug in the go-azure-sdk, in which case we shall open an issue towards it and attach the issue link here.

Reply:
updated the code to use subscription as the default scope when it's not specified by user